nemo_text_processing/text_normalization/de/taggers/electronic.py

```diff
@@ -72,7 +78,28 @@ def __init__(self, deterministic: bool = True):
         protocol = pynutil.insert('protocol: "') + protocol + pynutil.insert('"')
         url = protocol + insert_space + (domain_graph)
 
-        graph = url | domain_graph | email | tag
+        # Implements a graph for commonly-used hyphenated compound nouns (e.g. 3D-Drucker, 2D-Mammogram)
+        graph_abbreviation = pynini.string_file(get_abs_path("data/electronic/abbreviations.tsv"))
+        hyphen_accep = pynini.accep("-")
+        graph_compound_a = (
+            pynutil.insert("fragment_id:")
```

Comment:
is this valid for sparrowhawk?

Reply:
Yes. It's a valid field in SP's `Electronic` semiotic class implementation, but since it's better to move everything to Whitelist, then I suppose it doesn't matter?

Reply:
Mmm, what are our options with whitelist sparrowhawk wise?

Reply:
Seeing how the SP proto doesn't even list `Whitelist` as a semiotic class -- rightfully so IMO -- the options are endless? The current implementation of `Whitelist` doesn't do much besides case-folding and deterministic routing.  Moving the logic there should work, _however_ it will most definitely give precedence to cases that we may not want to prioritize (e.g. DGX-1 not being `ELECTRONIC` is probably more acceptable than `2-Liter` not being `MEASURE`).  On the other hand, leaving the current setup with the limitations imposed by SP's limitations lets you keep a tight lid on what "electronic" things go in it by simply editing the `.tsv` file.

Perhaps, the best of two worlds would be to leave the above in place, and expand `Whitelist` to match everything `ABC-### and `###-ABC`? That way all electronics stay `Electronic` and we have a `Whitelist` that captures `G-9`, `COVID-19`, etc.

Reply:
So to reiterate: we add the functionality to both whitelist and electronic, but whitelist is more general while elctronic captures specific whitelist?

I think that can work. But we can also just add negative weights within whitelist to take care of precedence cases.

Reply:
As _extra_ as that sounds -- yes. It may not matter for the final output, but I think it would be useful to separate evidently `ELECTRONIC` things from `WHITELIST` during tagging.

If you don't think it matters all that much (and it may not), I'll add the changes to `WHITELIST` and we push like that.

Reply:
Mmm, redundancy becomes a pain on debugging. Just move everything to whitelist, but make the key electronic terms in the actual whitelist while the  more liberal expansion a few lines in the tagger graph. Make the weight for the whitelist file have highest priority. If we get annoyed by the liberal expansion, we comment out and work on later. How does that sound?

Reply:
Sounds very sensible. Let me implement that and see how well it meshes with the rest of the the grammar.